library/cwm/src/lib/cwm/ui_state.rb

```diff
@@ -0,0 +1,209 @@
+# encoding: utf-8
+
+# Copyright (c) [2018] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "abstract_method"
+
+module CWM
+  # Singleton class to keep the position of the user in the UI and other similar
+  # information that needs to be rememberd across UI redraws to give the user a
+  # sense of continuity.
+  #
+  # If you want to see it in action, have a look at yast2-storage-ng or yast2-firewall
+  # modules.
+  #
+  # @example Defining a UIState class to handle firewall zones (extracted from yast2-firewall)
+  #
+  #   require "cwm/ui_state"
+  #   class MyUIState < CWM::UIState
+  #     def go_to_tree_node(page)
+  #       super
+  #       self.candidate_notes =
+  #         if page.respond_to?(:zone)
+  #           zone_page_candidates(page)
+  #         else
+  #           [page.label]
+  #         end
+  #       end
+  #     end
+  #
+  #   private
+  #
+  #     def zone_page_candidates(page)
+  #       [page.zone.name]
+  #     end
+  #   end
+  #
+  # @example Using the UIState from an CWM::TreePager
+  #
+  #   class OverviewTreePager < CWM::TreePager
+  #     # Overrides default behaviour to register the new state.
+  #     def switch_page(page)
+  #       MyUIState.instance.go_to_tree_node(page)
+  #       super
+  #     end
+  #
+  #     # Ensures the tree is properly initialized according to the UI state after
+  #     # a redraw
+  #     def initial_page
+  #       MyUIState.instance.find_tree_node(@pages) || super
+  #     end
+  #   end
+  #
+  # @example Registering the selected row from a CWM::Table
+  #
+  #   class MyTable < ::CWM::Table
+  #     def opt
+  #       [:notify, :immediate]
+  #     end
+  #
+  #     def handle(event)
+  #       MyUIState.instance.select_row(value) if event["EventReason"] == "SelectionChanged"
+  #       nil
+  #     end
+  #   end
+  #
+  class UIState
+    include Yast::I18n
+
+    # Constructor
+    #
+    # Called through {.create_instance}, starts with a blank situation (which
+    # means default for each widget will be honored).
+    def initialize
+      textdomain(textdomain_name)
```

Comment:
what is a goal of this call? I do not see any translatable string.

Reply:
It was in the original implementation, so I've decided to keep it. I will double check if it is still needed.

Reply:
My problem with it is that it just adds requirements ( defining textdomain name) without any benefits visible for me.